pulsar-client-cpp/lib/ConsumerImpl.cc

```diff
@@ -301,14 +315,33 @@ void ConsumerImpl::messageReceived(const ClientConnectionPtr& cnx, const proto::
 }
 
 // Zero Queue size is not supported with Batch Messages
-unsigned int ConsumerImpl::receiveIndividualMessagesFromBatch(Message& batchedMessage) {
+uint32_t ConsumerImpl::receiveIndividualMessagesFromBatch(const ClientConnectionPtr& cnx, Message& batchedMessage) {
     unsigned int batchSize = batchedMessage.impl_->metadata.num_messages_in_batch();
     batchAcknowledgementTracker_.receivedMessage(batchedMessage);
     LOG_DEBUG("Received Batch messages of size - " << batchSize);
-    for (int i=0; i<batchSize; i++) {
+
+    for (int i = 0; i < batchSize; i++) {
         batchedMessage.impl_->messageId.batchIndex_ = i;
         // This is a cheap copy since message contains only one shared pointer (impl_)
-        incomingMessages_.push(Commands::deSerializeSingleMessageInBatch(batchedMessage));
+        Message msg = Commands::deSerializeSingleMessageInBatch(batchedMessage);
+
+        if (startMessageId_.is_present()) {
+            const BatchMessageId& msgId = static_cast<const BatchMessageId&>(msg.getMessageId());
+
+            // If we are receiving a batch message, we need to discard messages that were prior
+            // to the startMessageId
+            if (msgId.ledgerId_ == startMessageId_.value().ledgerId_
+                    && msgId.entryId_ == startMessageId_.value().entryId_
+                    && msgId.batchIndex_ <= startMessageId_.value().batchIndex_) {
+                LOG_DEBUG(
+                        getName() << "Ignoring message from before the startMessageId" << msg.getMessageId());
+                increaseAvailablePermits(cnx);
+                continue;
+            }
+        }
+
+        // Regular path, append individual message to incoming messages queue
+        incomingMessages_.push(msg);
     }
     return batchSize;
```

Comment:
We shouldn't be returning the batchSize since not all messages made it to incomingMessages_
Just return the number of messages added to the incomingMessages_ queue

Reply:
Good point. I'll fix it.